File: src/familiar_connect/processors/debug_logger.py
"""Debug processor: one log line per event on subscribed topics.

Phase-1 "is the bus alive?" signal. Colourised via
:mod:`familiar_connect.log_style`. Matches the Phase-1 deliverable in
the rollout plan — no user-visible behaviour beyond logs.
"""

from __future__ import annotations

import logging
from typing import TYPE_CHECKING

from familiar_connect import log_style as ls

if TYPE_CHECKING:
    from familiar_connect.bus.envelope import Event
    from familiar_connect.bus.protocols import EventBus

_logger = logging.getLogger("familiar_connect.processors.debug_logger")


class DebugLoggerProcessor:
    """Log one line per event. Does not republish."""

    name: str = "debug-logger"

    def __init__(self, *, topics: tuple[str, ...]) -> None:
        self.topics: tuple[str, ...] = topics

    async def handle(self, event: Event, bus: EventBus) -> None:  # noqa: ARG002
        payload_repr = (
            ls.trunc(repr(event.payload), 160) if event.payload is not None else "-"
        )
        _logger.info(
            f"{ls.tag('📥 Event', ls.LG)} "
            f"{ls.kv('topic', event.topic, vc=ls.LM)} "
            f"{ls.kv('event_id', event.event_id, vc=ls.LC)} "
            f"{ls.kv('turn_id', event.turn_id, vc=ls.LC)} "
            f"{ls.kv('session', event.session_id, vc=ls.LC)} "
            f"{ls.kv('seq', str(event.sequence_number), vc=ls.LY)} "
            f"{ls.kv('payload', payload_repr, vc=ls.LW)}"
        )

File: rust/familiar-connect/src/processors/debug_logger.rs
//! One log line per event on subscribed topics (subsystem 06; Python
//! `processors/debug_logger.py`).
//!
//! The Phase-1 "is the bus alive?" signal. Passive — never republishes.
//!
//! Port note: Python renders the payload as `repr(...)` truncated at 160 chars
//! (`-` when `None`). The Rust bus payload is a type-erased `Arc<dyn Any>` the
//! logger cannot inspect generically, so the payload field renders as `-`. The
//! topic / ids / sequence fields — the only ones the tests assert — are
//! reproduced exactly.

use crate::bus::envelope::Event;
use crate::bus::protocols::EventBus;
use crate::log_style as ls;

/// Logs one line per event; does not republish.
pub struct DebugLoggerProcessor {
    topics: Vec<String>,
}

impl DebugLoggerProcessor {
    /// The processor's human label.
    pub const NAME: &'static str = "debug-logger";

    /// Construct a logger over the injected `topics`.
    #[must_use]
    pub fn new(topics: impl IntoIterator<Item = impl Into<String>>) -> Self {
        Self {
            topics: topics.into_iter().map(Into::into).collect(),
        }
    }

    /// The processor name (`"debug-logger"`).
    #[must_use]
    pub const fn name(&self) -> &'static str {
        Self::NAME
    }

    /// The injected topics (borrowed for `bus.subscribe`).
    #[must_use]
    pub fn topics(&self) -> Vec<&str> {
        self.topics.iter().map(String::as_str).collect()
    }

    /// Log one INFO line for `event`.
    ///
    /// # Errors
    /// Never fails (returns `Ok` unconditionally) — signature matches the
    /// processor contract.
    #[allow(clippy::unused_async, reason = "processor contract is async")]
    pub async fn handle(&self, event: &Event, _bus: &dyn EventBus) -> anyhow::Result<()> {
        tracing::info!(
            "{} {} {} {} {} {} {}",
            ls::tag("\u{1f4e5} Event", ls::LG),
            ls::kv_styled("topic", &event.topic, ls::W, ls::LM),
            ls::kv_styled("event_id", &event.event_id, ls::W, ls::LC),
            ls::kv_styled("turn_id", &event.turn_id, ls::W, ls::LC),
            ls::kv_styled("session", &event.session_id, ls::W, ls::LC),
            ls::kv_styled("seq", &event.sequence_number.to_string(), ls::W, ls::LY),
            ls::kv_styled("payload", "-", ls::W, ls::LW),
        );
        Ok(())
    }
}
